File: Unity/TSProj/src/mods/ModTheSpire/Scripts/DungeonManager.ts
import QueueMessageKit from "../../../Core/Module/Event/QueueMessageKit";
import { ObserverMessageKit } from "../../../Core/ObserverMessageKit";
import { AbstractBlight } from "./Blight/AbstractBlight";
import { AbstractDungeon } from "./Dungeon/AbstractDungeon";
import EffectKit from "./Effect/EffectKit";
import { AbstractRoom } from "./Room/AbstractRoom";
import { AbstractPlayer } from "./Unit/Character/AbstractPlayer";
import { PlayerInfo } from "./Unit/Character/PlayerInfo";

//DungeonManager类似于一次游戏的管理器.

//每次开始游戏都会重置局部变量.
export default class DungeonManager {
    //region 静态变量
    public static Inst: DungeonManager;
    public static ActionManager = new QueueMessageKit();
    public static MessageManager = new ObserverMessageKit();
    public static EffectManager = new EffectKit();
    //endregion
    
    //region局部变量
    //本局游戏操作的角色
    public Player : AbstractPlayer;
    //本局游戏爬塔楼层
    public FloorNum = 0;
    //灾祸池
    public BlightPool :Array<AbstractBlight>
    //本局游戏种子
    public Seed:string;
    //进阶等级
    public AdvanceLevel:number = 0;
    //是否是无尽模式
    private _IsEndless : boolean;
    //当前房间
    private _CurrentRoom:AbstractRoom;
    //当前章节地牢
    private _CurrentDungeon : AbstractDungeon;
    get CurrentDungeon(): AbstractDungeon {
        return this._CurrentDungeon;
    }

    set CurrentDungeon(value: AbstractDungeon) {
        if(this._CurrentDungeon != null)
            this._CurrentDungeon.Dispose();
        this._CurrentDungeon = value;
        this._CurrentDungeon.Init();
    }

    get CurrentRoom(): AbstractRoom {
        return this._CurrentRoom;
    }

    set CurrentRoom(value: AbstractRoom) {
        this._CurrentRoom = value;
        this._CurrentRoom.Enter();
    }

    get IsEndless(): boolean {
        return this._IsEndless;
    }

    set IsEndless(value: boolean) {
        if(this.FloorNum == 0)
            this._IsEndless = value;
    }
    //endregion
    
    //开始一局新的游戏
    public static NewGame(player:PlayerInfo, seed:string){
        let manager = new DungeonManager();
        manager.Player = new AbstractPlayer(player);
        manager.Seed = seed;
        manager.FloorNum = 0;
        this.Inst = manager;
    }
    
    public Update(){
        if(this.Player)
            this.Player.Update()
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Dungeon/DungeonManager.cs
using System.Collections.Generic;
using ModTheSpire.Blight;
using ModTheSpire.Effect;
using ModTheSpire.Event;
using ModTheSpire.Room;
using ModTheSpire.Unit.Character;

namespace ModTheSpire.Dungeon
{
//DungeonManager类似于一次游戏的管理器.

//每次开始游戏都会重置局部变量.
    public class DungeonManager
    {
        #region 静态变量

        public static DungeonManager Inst;
        public static QueueMessageKit ActionManager = QueueMessageKit.Inst;
        public static MessageKit MessageManager = MessageKit.Inst;
        public static EffectKit EffectManager = EffectKit.Inst;

        #endregion

        #region 局部变量

        //本局游戏操作的角色
        public AbstractPlayer Player;

        //本局游戏爬塔楼层
        public int FloorNum { get; set; }

        //灾祸池
        public List<AbstractBlight> BlightPool;

        //本局游戏种子
        public string Seed;

        //进阶等级
        public int AdvanceLevel = 0;

        //是否是无尽模式
        public bool IsEndless { get; set; }

        private AbstractRoom mCurrentRoom;

        //当前房间
        public AbstractRoom CurrentRoom
        {
            get => mCurrentRoom;
            set
            {
                mCurrentRoom = value;
                mCurrentRoom.Enter();
            }
        }

        private AbstractDungeon mCurrentDungeon;

        //当前章节地牢
        public AbstractDungeon CurrentDungeon
        {
            get => mCurrentDungeon;
            set
            {
                if (mCurrentDungeon != null)
                    mCurrentDungeon.Dispose();
                mCurrentDungeon = value;
                mCurrentDungeon.Init();
            }
        }

        #endregion

        //开始一局新的游戏
        public static void NewGame(IPlayerInfo player, string seed)
        {
            var manager = new DungeonManager
            {
                Player = new AbstractPlayer(player),
                Seed = seed,
                FloorNum = 0
            };
            Inst = manager;
        }

        public void Update()
        {
            if (this.Player != null)
                this.Player.Update();
        }
    }
}